File: minestom/src/main/java/org/readutf/engine/minestom/arena/MinestomArenaPlatform.java
package org.readutf.engine.minestom.arena;

import com.fasterxml.jackson.databind.ObjectMapper;
import net.hollowcube.polar.PolarLoader;
import net.minestom.server.MinecraftServer;
import net.minestom.server.instance.Instance;
import net.minestom.server.world.DimensionType;
import org.jetbrains.annotations.NotNull;
import org.readutf.buildformat.BuildData;
import org.readutf.buildformat.BuildMeta;
import org.readutf.buildformat.store.BuildDataStore;
import org.readutf.engine.arena.Arena;
import org.readutf.engine.arena.ArenaPlatform;
import org.readutf.engine.arena.build.BuildPlacement;
import org.readutf.engine.arena.exception.ArenaLoadException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.util.*;

public class MinestomArenaPlatform implements ArenaPlatform<Instance> {

    private static final Logger logger = LoggerFactory.getLogger(MinestomArenaPlatform.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private @NotNull
    final BuildDataStore buildDataStore;

    public MinestomArenaPlatform(@NotNull BuildDataStore buildDataStore, @NotNull File cacheDirector) {
        this.buildDataStore = buildDataStore;
    }

    @Override
    public @NotNull BuildPlacement<Instance> placeBuild(int buildId, @NotNull BuildMeta buildMeta) throws ArenaLoadException, Exception {

        BuildData buildData = buildDataStore.get(buildMeta.name(), buildMeta.version());
        if (buildData == null) throw new ArenaLoadException("Build not found: " + buildMeta.name());

        Instance instance = MinecraftServer.getInstanceManager().createInstanceContainer(
                DimensionType.OVERWORLD,
                new PolarLoader(new ByteArrayInputStream(buildData.polarData())));

        return new BuildPlacement<>(instance);
    }

    @Override
    public void freeArena(Arena<Instance, ?> arena) {
        MinecraftServer.getInstanceManager().unregisterInstance(arena.getWorld());
    }


}

File: minestom/src/main/kotlin/org/readutf/game/minestom/arena/MinestomArenaPlatform.kt
package org.readutf.game.minestom.arena

import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import net.hollowcube.schem.reader.SpongeSchematicReader
import net.kyori.adventure.text.Component
import net.minestom.server.MinecraftServer
import net.minestom.server.coordinate.Pos
import net.minestom.server.instance.Instance
import org.readutf.buildformat.common.Build
import org.readutf.buildformat.common.markers.Position
import org.readutf.game.engine.arena.Arena
import org.readutf.game.engine.arena.ArenaPlatform
import org.readutf.game.engine.arena.BuildPlacement
import org.readutf.game.minestom.platform.MinestomWorld
import java.util.UUID

class MinestomArenaPlatform : ArenaPlatform {

    private val instances = mutableMapOf<UUID, Instance>()

    override fun placeBuild(buildId: UUID, build: Build): Result<BuildPlacement, Throwable> {
        val instance = MinecraftServer.getInstanceManager().createInstanceContainer()

        val schematic = SpongeSchematicReader().read(build.buildSchematic.buildData)

        schematic.createBatch().apply(instance, Pos.ZERO) {
        }

        instances[buildId] = instance

        return Ok(
            BuildPlacement(
                world = MinestomWorld(instance),
                origin = Position.ZERO,
                markers = emptyList(),
            ),
        )
    }

    override fun freeArena(arena: Arena<*>): Result<Unit, Throwable> {
        instances[arena.buildId]?.let { instance ->
            instance.players.forEach { it.kick(Component.empty()) }

            MinecraftServer.getInstanceManager().unregisterInstance(instance)
        }
        return Ok(Unit)
    }
}
